module/zfs/vfs_ratelimit.c

```diff
@@ -0,0 +1,688 @@
+ * is done in the kernel and isn't shown in the properties.
+ *
+ * How does the accounting work?
+ *
+ * When a request comes, we may need to consider multiple limits.
+ * For example a data read request of eg. 192kB (with 128kB recordsize) is
+ * accounted as 192kB bandwidth read, 192kB bandwidth total, two operations read
+ * and two operations total. Not all of those limits have to be configured or
+ * some might be configured on a dataset and others on a parent dataset(s).
+ *
+ * For each type we use two fields to track the wait times: rl_timeslot and
+ * rl_reminder. rl_timeslot holds the point in time up to which the last
+ * processes is waiting for. If the rl_timeslot is lower than the current time,
+ * it means that no processes are waiting. rl_reminder is the amount of data
+ * modulo the limit. For example if we have a read bandwidth limit of 64MB/s,
+ * so it is 4MB per 1/16s. The process is trying to read 11MB. This would
+ * give us rl_timeslot = now + 2 (we account for 2 full time slots of 1/16s)
+ * and rl_reminder = 3MB. This process has to sleep for 2/16s. When immediately
+ * another process is trying to read 1MB, this 1MB will be added to the current
+ * rl_reminder giving 4MB, so full limit unit for 1/16s. Now rl_timeslot will
+ * be set to now + 3 and rl_reminder to 0. The last process is going to sleep
+ * for 3/16s.
+ */
+
+/*
+ * Number of slots we divide one second into. More granularity is better for
+ * interactivity, but for small limits we may lose some precision.
+ */
+#define	RATELIMIT_RESOLUTION	16
+
+struct vfs_ratelimit {
+	kmutex_t	rl_lock;
+	uint64_t	rl_limits[ZFS_RATELIMIT_NTYPES];
+	uint64_t	rl_timeslot[ZFS_RATELIMIT_NTYPES];
+	uint64_t	rl_reminder[ZFS_RATELIMIT_NTYPES];
+};
+
+int
+vfs_ratelimit_prop_to_type(zfs_prop_t prop)
+{
+
+	switch (prop) {
+	case ZFS_PROP_RATELIMIT_BW_READ:
+		return (ZFS_RATELIMIT_BW_READ);
+	case ZFS_PROP_RATELIMIT_BW_WRITE:
+		return (ZFS_RATELIMIT_BW_WRITE);
+	case ZFS_PROP_RATELIMIT_BW_TOTAL:
+		return (ZFS_RATELIMIT_BW_TOTAL);
+	case ZFS_PROP_RATELIMIT_OP_READ:
+		return (ZFS_RATELIMIT_OP_READ);
+	case ZFS_PROP_RATELIMIT_OP_WRITE:
+		return (ZFS_RATELIMIT_OP_WRITE);
+	case ZFS_PROP_RATELIMIT_OP_TOTAL:
+		return (ZFS_RATELIMIT_OP_TOTAL);
+	default:
+		panic("Invalid property %d", prop);
+	}
+}
+
+zfs_prop_t
+vfs_ratelimit_type_to_prop(int type)
+{
+
+	switch (type) {
+	case ZFS_RATELIMIT_BW_READ:
+		return (ZFS_PROP_RATELIMIT_BW_READ);
+	case ZFS_RATELIMIT_BW_WRITE:
+		return (ZFS_PROP_RATELIMIT_BW_WRITE);
+	case ZFS_RATELIMIT_BW_TOTAL:
+		return (ZFS_PROP_RATELIMIT_BW_TOTAL);
+	case ZFS_RATELIMIT_OP_READ:
+		return (ZFS_PROP_RATELIMIT_OP_READ);
+	case ZFS_RATELIMIT_OP_WRITE:
+		return (ZFS_PROP_RATELIMIT_OP_WRITE);
+	case ZFS_RATELIMIT_OP_TOTAL:
+		return (ZFS_PROP_RATELIMIT_OP_TOTAL);
+	default:
+		panic("Invalid type %d", type);
+	}
+}
+
+static boolean_t
+ratelimit_is_none(const uint64_t *limits)
+{
+
+	for (int i = ZFS_RATELIMIT_FIRST; i < ZFS_RATELIMIT_NTYPES; i++) {
```

Comment:
```suggestion
	for (int i = ZFS_RATELIMIT_FIRST; i <= ZFS_RATELIMIT_LAST; i++) {
```
The effect is the same, just IMHO using ZFS_RATELIMIT_NTYPES in combination with ZFS_RATELIMIT_FIRST is wrong.  If you prefer ZFS_RATELIMIT_NTYPES, then I'd start from 0.  Though both ways assume there are not holes.  The same is below in `vfs_ratelimit_alloc()`.